File: ios/MedicalFactChecker/Sources/Views/Components/FullTextSourceBadge.swift
#if os(iOS)
// BMLibrarian Lite - Biomedical Literature Research Tool
// Copyright (C) 2024-2026 Dr Horst Herb
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

import SwiftUI

// MARK: - Constants

/// Constants for the source badge UI.
private enum SourceBadgeConstants {
    /// Horizontal padding inside the badge.
    static let horizontalPadding: CGFloat = 6

    /// Vertical padding inside the badge.
    static let verticalPadding: CGFloat = 2

    /// Corner radius for the badge background.
    static let cornerRadius: CGFloat = 4

    /// Opacity for the background color.
    static let backgroundOpacity: Double = 0.15
}

// MARK: - Full Text Source Badge

/// Small badge showing where full text came from.
///
/// Displays an icon and text label indicating the source of the full text
/// (Europe PMC, Unpaywall, Publisher, or Cached).
struct FullTextSourceBadge: View {
    /// The source to display.
    let source: AppFullTextSource

    var body: some View {
        HStack(spacing: 4) {
            Image(systemName: source.iconName)
                .font(.caption2)
            Text(source.displayName)
                .font(.caption2)
        }
        .foregroundColor(badgeColor)
        .padding(.horizontal, SourceBadgeConstants.horizontalPadding)
        .padding(.vertical, SourceBadgeConstants.verticalPadding)
        .background(badgeColor.opacity(SourceBadgeConstants.backgroundOpacity))
        .cornerRadius(SourceBadgeConstants.cornerRadius)
        .accessibilityLabel("Full text from \(source.displayName)")
    }

    /// Color for the badge based on the source.
    private var badgeColor: Color {
        switch source {
        case .europePMC: return .blue
        case .unpaywall: return .green
        case .doi: return .orange
        case .cached: return .gray
        case .uploaded: return .purple
        }
    }
}

// MARK: - Preview

#Preview {
    VStack(spacing: 8) {
        FullTextSourceBadge(source: .europePMC)
        FullTextSourceBadge(source: .unpaywall)
        FullTextSourceBadge(source: .doi)
        FullTextSourceBadge(source: .cached)
        FullTextSourceBadge(source: .uploaded)
    }
    .padding()
}

#endif // os(iOS)

File: android/MedicalFactChecker/app/src/main/java/com/bmlibrarian/factchecker/ui/fulltext/components/FullTextSourceBadge.kt
/*
 * BMLibrarian Lite - Biomedical Literature Research Tool
 * Copyright (C) 2024-2025 Dr Horst Herb
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package com.bmlibrarian.factchecker.ui.fulltext.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import com.bmlibrarian.factchecker.util.Constants

/**
 * Badge displaying the source of full-text content.
 *
 * @param source Source name (e.g., "Europe PMC", "Unpaywall", "Publisher").
 * @param modifier Modifier for the badge.
 */
@Composable
fun FullTextSourceBadge(
    source: String,
    modifier: Modifier = Modifier
) {
    val (backgroundColor, textColor) = when (source.lowercase()) {
        "europe pmc" -> Pair(Color(0xFF1976D2), Color.White)
        "unpaywall" -> Pair(Color(0xFF4CAF50), Color.White)
        "publisher", "doi" -> Pair(Color(0xFF9E9E9E), Color.White)
        "cached" -> Pair(Color(0xFF607D8B), Color.White)
        else -> Pair(MaterialTheme.colorScheme.surfaceVariant, MaterialTheme.colorScheme.onSurfaceVariant)
    }

    Text(
        text = source,
        style = MaterialTheme.typography.labelSmall,
        color = textColor,
        modifier = modifier
            .background(
                color = backgroundColor,
                shape = RoundedCornerShape(4.dp)
            )
            .padding(
                horizontal = Constants.UI_BADGE_PADDING_HORIZONTAL.dp,
                vertical = Constants.UI_BADGE_PADDING_VERTICAL.dp
            )
    )
}
